src/Type/Php/ArrayChunkFunctionReturnTypeExtension.php

```diff
@@ -46,22 +48,32 @@ public function getTypeFromFunctionCall(FunctionReflection $functionReflection,
 			$preserveKeys = false;
 		}
 
-		if ($lengthType instanceof ConstantIntegerType && $lengthType->getValue() < 1) {
+		$negativeOrZero = IntegerRangeType::fromInterval(null, 0);
+		if ($negativeOrZero->isSuperTypeOf($lengthType)->yes()) {
 			return $this->phpVersion->throwsValueErrorForInternalFunctions() ? new NeverType() : new NullType();
 		}
 
 		if (!$arrayType->isArray()->yes()) {
 			return null;
 		}
 
-		$constantArrays = $arrayType->getConstantArrays();
-		if ($lengthType instanceof ConstantIntegerType && $lengthType->getValue() >= 1 && $preserveKeys !== null && count($constantArrays) > 0) {
-			$results = [];
-			foreach ($constantArrays as $constantArray) {
-				$results[] = $constantArray->chunk($lengthType->getValue(), $preserveKeys);
+		if ($preserveKeys !== null) {
+			$constantArrays = $arrayType->getConstantArrays();
+			$biggerOne = IntegerRangeType::fromInterval(1, null);
+			if (count($constantArrays) > 0 && $biggerOne->isSuperTypeOf($lengthType)->yes()) {
+				$results = [];
+				foreach ($constantArrays as $constantArray) {
+					foreach ($lengthType->getFiniteTypes() as $finiteType) {
+						if (!$finiteType instanceof ConstantIntegerType || $finiteType->getValue() < 1) {
```

Comment:
This is a courageous assumption. I wouldn't throw `ShouldNotHappenException`, it can break right after https://github.com/phpstan/phpstan-src/pull/2642/files gets merged.

So instead it'd be better to return the default return type, instead of crashing.

Also, I'd limit the possibilities to a small number of finite types, rather than have a performance problem on our hands. It can get rather big.

Also - you need to test this with `$lengthType` of infinite interval like `int<1, max>` or `int<50, max>`.